File: leetcode/0525.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> mp;
        mp[0] = -1;
        int presum = 0, n = nums.size(), ans = 0;
        for (int i = 0; i < n; ++i) {
            if (nums[i])
                presum++;
            else
                presum--;
            if (mp.find(presum) != mp.end()) {
                ans = max(ans, i - mp[presum]);
            } else {
                mp[presum] = i;
            }
        }
        return ans;
    }
};

File: leetcode/0525.go
package leetcode

func findMaxLength(nums []int) int {
	mp := map[int]int{0: -1}
	presum, ans := 0, 0
	for i, num := range nums {
		if num == 1 {
			presum++
		} else {
			presum--
		}
		if index, ok := mp[presum]; ok {
			if ans < i-index {
				ans = i - index
			}
		} else {
			mp[presum] = i
		}
	}
	return ans
}
